session/session.go

```diff
@@ -3496,10 +3497,44 @@ func (s *session) GetStmtStats() *stmtstats.StatementStats {
 
 // EncodeSessionStates implements SessionStatesHandler.EncodeSessionStates interface.
 func (s *session) EncodeSessionStates(ctx context.Context, sctx sessionctx.Context, sessionStates *sessionstates.SessionStates) (err error) {
-	return s.sessionVars.EncodeSessionStates(ctx, sessionStates)
+	if err = s.sessionVars.EncodeSessionStates(ctx, sessionStates); err != nil {
+		return err
+	}
+
+	// Encode session variables. We put it here instead of SessionVars to avoid cycle import.
+	sessionStates.SystemVars = make(map[string]string)
+	for _, sv := range variable.GetSysVars() {
+		switch {
+		case sv.Hidden, sv.HasNoneScope(), sv.HasInstanceScope(), !sv.HasSessionScope():
+			// Hidden, none-scoped, and instance-scoped variables cannot be modified.
```

Comment:
> instance-scoped variables cannot be modified

Instance-scoped variables could be modified even at runtime, but I agree with you that instance-scoped variables should not be encoded and restored.

Please update the comment here.